allennlp/nn/beam_search.py

```diff
@@ -489,7 +546,34 @@ def top_p_sampling(
             sampler=sampler_p,
         )
 
+    @classmethod
+    def stochastic_beam_search(
+        cls,
+        end_index: int,
+        max_steps: int = 50,
+        beam_size: int = 10,
+        per_node_beam_size: int = None,
+    ) -> "BeamSearch":
+        """
+        Given an index of the end token in target vocabulary, return a `BeamSearch` object
+        that can be used to find `beam_size` candidate sequences, found by using Stochastic
+        Beam Search, which leverages Gumbel-TopK sampling. See
+        [*Stochastic Beams and Where to Find Them: The Gumbel-Top-k Trick for Sampling
+        Sequences Without Replacement*, W Kool, H Van Hoof and M Welling, 2010]
+        (https://arxiv.org/abs/1903.06059).
+        """
+
+        gumbel_sampler = GumbelMaxSampler()
+        return cls(
+            end_index=end_index,
+            max_steps=max_steps,
+            beam_size=beam_size,
+            per_node_beam_size=per_node_beam_size or beam_size,
+            sampler=gumbel_sampler,
+        )
+
 
 BeamSearch.register("without_sampling", constructor="without_sampling")(BeamSearch)
 BeamSearch.register("top_p_sampling", constructor="top_p_sampling")(BeamSearch)
 BeamSearch.register("top_k_sampling", constructor="top_k_sampling")(BeamSearch)
+BeamSearch.register("stochastic_beam_search", constructor="stochastic_beam_search")(BeamSearch)
```

Comment:
Same suggestion here ⬇️

```suggestion
BeamSearch.register("stochastic", constructor="stochastic")(BeamSearch)
```